globus_sdk/auth/identity_map.py

```diff
@@ -0,0 +1,222 @@
+import re
+
+# what qualifies as a valid Identity Username?
+_USERNAME_PATTERN = re.compile(r"^[a-zA-Z0-9]+.*@[a-zA-z0-9-]+\..*[a-zA-Z]+$")
+
+
+def is_username(val):
+    return _USERNAME_PATTERN.match(val) is not None
+
+
+def split_ids_and_usernames(identity_ids):
+    ids = set()
+    usernames = set()
+
+    for val in identity_ids:
+        if is_username(val):
+            usernames.add(val)
+        else:
+            ids.add(val)
+
+    return ids, usernames
+
+
+class IdentityMap(object):
+    r"""
+    There's a common pattern of having a large batch of Globus Auth Identities which you
+    want to inspect. For example, you may have a list of identity IDs fetched from
+    Access Control Lists on Globus Endpoints. In order to display these identities to an
+    end user, you may want to resolve them to usernames.
+
+    However, naively looking up the identities one-by-one is very inefficient. It's best
+    to do batched lookups with multiple identities at once. In these cases, an
+    ``IdentityMap`` can be used to do those batched lookups for you.
+
+    An ``IdentityMap`` is a mapping-like type which converts Identity IDs and Identity
+    Names to Identity records (dictionaries) using the Globus Auth API.
+
+    .. note::
+
+        ``IdentityMap`` objects are not full Mappings in the same sense as python dicts
+        and similar objects. By design, they only implement a small part of the Mapping
+        protocol.
+
+    The basic usage pattern is
+
+    - create an ``IdentityMap`` with an AuthClient which will be used to call out to
+      Globus Auth
+
+    - seed the ``IdentityMap`` with IDs and Usernames via :py:meth:`~IdentityMap.add` (you
+      can also do this during initialization)
+
+    - retreive identity IDs or Usernames from the map
+
+    Because the map can be populated with a collection of identity IDs and Usernames
+    prior to lookups being performed, it can improve the efficiency of these operations
+    up to 100x over individual lookups.
+
+    If you attempt to retrieve an identity which has not been previously added to the
+    map, it will be immediately added. But adding many identities beforehand will
+    improve performance.
+
+    The ``IdentityMap`` will cache its results so that repeated lookups of the same Identity
+    will not repeat work. It will also map identities both by ID and by Username,
+    regardless of how they're initially looked up.
+
+    .. warning::
+
+        If an Identity is not found in Globus Auth, it will trigger a KeyError when
+        looked up. Your code must be ready to handle KeyErrors when doing a lookup.
+
+    Correct usage looks something like so::
+
+        ac = globus_sdk.AuthClient(...)
+        idmap = globus_sdk.ext.IdentityMap(
```

Comment:
This isn't in an `ext` namespace anymore, right?

Reply:
Good catch, thank you. Yes, per our brief chat on that subject, no more `globus_sdk.ext` subpackage. I'll grep for other cases of it.